File: core/src/main/scala/pl/edu/agh/multiscalemodelling/engine/logic/Logic.java
package pl.edu.agh.multiscalemodelling.engine.logic;

public abstract class Logic {

    protected Board board;
    protected boolean isPaused;

    public Logic() {
        isPaused = true;
    }


    public void pause() {

        isPaused = !isPaused;

    }

    public void iterate() {

        //long startTime = System.currentTimeMillis();
        int threadsNum = Runtime.getRuntime().availableProcessors() * Runtime.getRuntime().availableProcessors();
        //threadsNum = 1;
        int start = 0;
        int end = 0;
        Thread[] threads = new Thread[threadsNum];

        //board.setNeighbourhood(new PentagonalRandomNeighbourhood(), new PeriodicBoudaryCondition());

        for (int i = 0; i < threadsNum; i++) {

            if (board.cells.size() % threadsNum == 0) {

                end += board.cells.size() / threadsNum;

            } else {

                end += (board.cells.size() / threadsNum) + 1;

            }

            int warunek = (end - start) * (i + 1);

            if (warunek > board.cells.size()) {
                end -= (warunek - board.cells.size());
            }

            threads[i] = createThread(board, start, end);
            threads[i].setName(Integer.toString(i));
            threads[i].start();

            start = end;

        }

        for (int i = 0; i < threadsNum; i++) {

            try {
                threads[i].join();
            } catch (InterruptedException e) {
            }

        }

        long endTime = System.currentTimeMillis();

        //System.out.println((endTime - startTime));


    }

    public Thread createThread(Board board, int start, int end) {

        return new LogicThread(board.cells.subList(start, end));

    }

    public boolean isPaused() {

        return isPaused;

    }

    public abstract void click(int x, int y);

    public Board getBoard() {
        return board;
    }

    public void setBoard(Board board) {
        this.board = board;
    }
}

File: core/src/main/scala/pl/edu/agh/multiscalemodelling/engine/logic/Logic.scala
package pl.edu.agh.multiscalemodelling.engine.logic

object Logic {
  var operationMode: OperationMode.Value = OperationMode.SIMPLE_GROWTH
  var allProcessed = false
}

abstract class Logic() {

  var board: Board = _
  var isPaused = false
  var started = false

  isPaused = true
  started = false

  def pause(): Unit = isPaused = !isPaused

  def iterate(): Unit = {
    Logic.allProcessed = true
    //val threadsNum = Runtime.getRuntime.availableProcessors * Runtime.getRuntime.availableProcessors
    val threadsNum = 1
    var start = 0
    var end = 0
    val threads = new Array[Thread](threadsNum)
    var i = 0
    while (i < threadsNum) {
      if (board.cells.size % threadsNum == 0) end += board.cells.size / threadsNum
      else end += (board.cells.size / threadsNum) + 1
      val condition = (end - start) * (i + 1)
      if (condition > board.cells.size) end -= (condition - board.cells.size)
      threads(i) = Logic.operationMode match {
        case OperationMode.SIMPLE_GROWTH => createThread(board, start, end)
        case OperationMode.SIMPLE_MCS => new MonteCarloLogicThread(board.cells.subList(start, end))
      }
      threads(i).setName(Integer.toString(i))
      threads(i).start()

      start = end
      i += 1
      i - 1
    }
    i = 0
    while ( {
      i < threadsNum
    }) {
      try threads(i).join()
      catch {
        case e: InterruptedException =>
      }
      i += 1
      i - 1
    }
  }

  def createThread(board: Board, start: Int, end: Int) = new LogicThread(board.cells.subList(start, end))

  def click(x: Int, y: Int): Unit

  def getBoard: Board = board

  def setBoard(board: Board): Unit = this.board = board
}